source-code/core/src/plugin/setupPlugins.ts

```diff
@@ -0,0 +1,39 @@
+import type { Config, EnvironmentFunctions } from "../config/schema.js"
+import type { Plugin, PluginSetupFunction } from "./types.js"
+import { deepmergeInto } from "deepmerge-ts"
+
+export type ConfigWithSetupPlugins = Omit<Partial<Config>, "plugins"> & {
+	plugins: Plugin[]
+}
+
+/**
+ * Setup the plugins and process the config callback.
+ *
+ * The function mutates the config object. Mutating the config object
```

Comment:
which performance benefits?
I would use `structuredClone` before line 22 to make a copy of the `config`

Reply:
Structured clone makes no sense in this context. 

This code is about merging two objects, not copying and returning a modified copy. If we want immutability, we need to change line 34 from `deepmergeInto` to `deepmerge`. 

But 1) immutability doesn't seem useful because the config is not mutated until a plugin's config is known to work and 2) makes the API harder to consumer, see https://github.com/inlang/inlang/pull/605/files/a717f2296ee4b3a91d012ade6fdfeea64b48048a#diff-b9bda77260de9836d5a066f2123040a864c1f309127aecd6fd8eb288c20369b1R10.

Reply:
Added more context and error handling (to not corrupt the config when one plugins crashes) ec15170

Reply:
> Initial load of the inlang editor when a project uses 10+ plugins. I'll add that remark.

It would be always the same performance hit, because the clone get's perfomed a single time before executing all plugins.

> 1) immutability doesn't seem useful because the config is not mutated until a plugin's config is known to work

We are passing the reference to the holy grail (inlang config) to all plugins. They can do whatever they want. They can call `delete config.referenceLanguage` and other stuff that could break.
In general I prefer the pure function approach, because it is easier to find bugs if functions don't have side effects.

Another poin that came to my mind: we should probably check if the config is valid after each plugin get's executed to have a better debugging experience.

> 2) makes the API harder to consume

Common, it only get'S executed a single time at a single place
`await setupPlugins({ config, env: args.env })`
vs.
`const configWithPlugins = await setupPlugins({ config, env: args.env })`
^^

Reply:
There seems to be a misconception:

Configs that plugins define are immutable. The merger of the returned configs of plugins is mutable.

> We are passing the reference to the holy grail (inlang config) to all plugins. 

We don't. 

Plugins create their own config object. 

https://github.com/inlang/inlang/blob/4babe4729d05d9fcf858a50e9d5c0ab9e4bdf452/source-code/core/src/plugin/createPlugin.test.ts#L13-L16

> Another poin that came to my mind: we should probably check if the config is valid after each plugin get's executed to have a better debugging experience.

Expensive. The test should occur in testing the plugin rather than on runtime when the editor loads. I created an issue https://github.com/inlang/inlang/issues/606.  

>> Initial load of the inlang editor when a project uses 10+ plugins. I'll add that remark.
> It would be always the same performance hit, because the clone get's perfomed a single time before executing all plugins.

If the intention is to disallow plugins to "break" the config (which can't happen to my knowledge), we would need to clone the object for each plugin. Otherwise, the entire inlang crashes because one plugin screws up the config. Subsequently, the big O notation is O(n), not O(1). The mutable approach is O(0) (does O(0) exist?) because we avoid cloning objects in the merger of all configs altogether. 

This is an internal design decision that we can easily change as we please. Thus, let's not argue too much and see what works and what does not. Right now, the immutable approach has a performance downside with no (clear) upside.

Reply:
Ok, let's wait until people complain 👍